Topic/move datetime validation

wolfsage opened the pull request.

wolfsage committed:
Move datetime field checks to _ix_check_user_properties.

This means datetime fields are now checked for validity on update, where
previously they were only checked on create.

The rules are now:

 * Inputs to these fields must be in RFC-3999 "zulu" format:
   ex: 2016-01-01T12:34:57Z

 * undef is allowed only if the field is optional

 * DateTime objects are allowed as inputs

wolfsage committed:
datetime fields automatically become timestamptz in Pg.

Also, datetime fields are now returned in RFC3999 "Zulu" format.

wolfsage commented on lib/Ix/DBIC/ResultSet.pm:
```diff
@@ -471,6 +449,29 @@ sub _ix_check_user_properties (
+        }
+      }
+    }
+
```
Perhaps the above block should be reordered so any valid cases fall through for other tests that may be added (instead of next-ing out on line 456)?

rjbs replied:
I think there's a bigger but related issue.  With your change, there are two places where we can skip the a field for being undefined: once at line 459 if it's an undefined where we expect a mandatory date, and once at 487 when it's the generic check used for undefined mandatory properties.

This seems like needless duplication.  We only need to check the definedness once, when we check it for everything else.

Then I wonder whether the datetime check shouldn't go before the validator.  Since it normalizes input strings into objects, then validator checks (like `$info->{validator}` stores) could be written expecting DateTime objects, so you could write a validator to only allow dates in the future.

Does that make sense?

wolfsage replied:
Yes, will try to rework this.

rjbs commented on lib/Ix/DBIC/ResultSet.pm:
```diff
@@ -515,6 +516,15 @@ sub _ix_wash_rows ($self, $rows) {
+          $row->{$key} = parsepgdate($row->{$key});
+        }
+      }
+    }
```
So, this code (wash_rows) runs on data that we're going to turn return in a method result payload.  It does stuff like, as above, turn `1` in a boolean column into a `true`.  I think what we want is to make sure that any datetime values are Ix::DateTime objects.  Then, perl consumers of results are guaranteed to have objects and JSON will get the formatted data.

Does this check really account for the possible things in a row?  That is: this code seems predicated on the idea that any value in `$row->{datetime_field}` is either undef, a 3339 string, or a pgsql string.  Are those the right options?

I'm not sure, I just want to make sure you thought it through.  Something about this code makes me go "hm."

wolfsage replied:
This method seems to get data from three places -
- User input
- Defaults
- The database

User input is always turned into an Ix::DateTime object, so we can trust that to look right (I think).

Defaults, now that I look at them more, could present a problem. They aren't validated against _ix_check_user_properties, so something unexpected could be in there. What do we do about that? (For example, the string "now()" is perfectly valid as far as the database is concerned!)

Since the database is storing a timestamptz, we _should_ always get something back that DateTime::Format::Pg should understand, (or null which we ignore), so that should be fine.

rjbs replied:
I thought about the defaults last night, too.  I think that for now we should assume that defaults are always valid. In the future, maybe we'll want to validate them, and that will require some more significant rejiggering, but for now we can just say "don't screw up your own defaults, programmer" and I think it's okay. Agreed?

rjbs commented on t/basic.t:
```diff
@@ -638,4 +638,160 @@ subtest "duplicated creation ids" => sub {
+            green => superhashof({
+              type => 'invalidProperties',
+              propertyErrors => { expires_at => 'no value given for required field' },
+            }),
```
I am going to be a **monster** and move the goal posts here.

In fact, you did give a value: you gave `null`.  If the field had been missing, it would've gotten a default.  Do you think we can/should give a better error message here?  I bet we can and should.  Something like "null value given but [type] required".  Legit?

wolfsage replied:
Yeah, good point.

rjbs commented:
This looks pretty good!  I made some comments, though, on things that I think should be reworked.  Let me know what you think about my thinking.

wolfsage committed:
Validate ix_default_properties as well.

This should provide some sanity to consumers of Ix.

wolfsage committed:
Don't duplicate checks for required fields.

wolfsage committed:
Provide more accurate error message when null is provided for req'd field.

wolfsage committed:
Remain 'user_prop' to 'properties' as it can contain defaults too.

It really means "All properties we are about to set" now.

wolfsage commented on lib/Ix/DBIC/ResultSet.pm:
```diff
@@ -395,7 +368,7 @@ sub ix_create ($self, $ctx, $to_create) {
 
     if ($row) {
-      my @defaults = grep {; ! exists $user_prop->{$_} } keys %default_properties;
+      my @defaults = grep {; ! exists $this->{$_} } keys %default_properties;
```
We can't check "$user_prop" (now $properties) anymore because it contains defaults as well, so we must look at the original input.

wolfsage committed:
Simplify processing of defaults and user inputs.

Also, ensure DateTime objects are only allowed for datetime fields.

wolfsage committed:
Add tests to ensure we handle objects as inputs correctly.

wolfsage committed:
Anchor beginning of date regex during testing.

wolfsage committed:
Don't depend on state being static. Previous tests could change it.

wolfsage committed:
Check defined before comparing against regex to avoid warning.

Test::Deep will throw a warning like:

  Use of uninitialized value $got in pattern match (m//) at
  /Test/Deep/Regexp.pm line 57.

if it tries to compare a similar looking structure to an expected
structure with a regexp.

Consider:

  use Test::Deep;
  use Test::More;

  cmp_deeply(
    [ { a => undef }, { a => 1 } ],
    set(
     { a => re(qr/\d/) }, { a => undef },
    ),
    "Ok!"
  );

  done_testing;

rjbs commented:
Merged!
